src/main/java/com/unstoppabledomains/resolution/ENS.java

```diff
@@ -0,0 +1,97 @@
+package com.unstoppabledomains.resolution;
+
+import java.util.Arrays;
+
+import com.unstoppabledomains.exceptions.NSExceptionCode;
+import com.unstoppabledomains.exceptions.NSExceptionParams;
+import com.unstoppabledomains.exceptions.NamingServiceException;
+import com.unstoppabledomains.resolution.artifacts.Numeric;
+import com.unstoppabledomains.resolution.contracts.Contract;
+import com.unstoppabledomains.resolution.contracts.ens.EnsContractType;
+import com.unstoppabledomains.resolution.contracts.ens.Registry;
+import com.unstoppabledomains.resolution.contracts.ens.Resolver;
+
+public class ENS extends NamingService {
+
+  static final String REGISTRY_ADDRESS = "0x00000000000C2E074eC69A0dFb2997BA6C7d2e1e";
+  private final String providerURL;
+  private final Registry registryContract;
+  
+
+  public ENS(String blockchainProviderURL) {
+    super();
+    this.providerURL = blockchainProviderURL;
+    this.registryContract = (Registry)buildContract(REGISTRY_ADDRESS, EnsContractType.Registry);
+  }
+
+  @Override
+  Boolean isSupported(String domain) {
+    String[] ensTLDs = { "eth", "kred", "luxe", "xyz" };
+    String[] split = domain.split("\\.");
+    String tld = split[split.length - 1];
+    return (split.length != 0 && Arrays.asList(ensTLDs).contains(tld));
+  }
+
+  @Override
+  String addr(String domain, String ticker) throws NamingServiceException {
+    if (!ticker.equalsIgnoreCase("ETH")) {
+      throw new NamingServiceException(NSExceptionCode.UnsupportedCurrency, new NSExceptionParams("c", ticker.toUpperCase()));
+    }
+    Resolver resolver = getResolverContract(domain);
+    byte[] tokenId = this.tokenId(domain);
+    return resolver.addr(tokenId, ticker.toUpperCase());
+  }
+
+  @Override
+  String email(String domain) throws NamingServiceException {
+    Resolver resolver = getResolverContract(domain);
+    byte[] tokenId = this.tokenId(domain);
+    String emailRecord = resolver.getTextRecord(tokenId, "whois.email.value");
+    if (Utilities.isNull(emailRecord)) {
+      throw new NamingServiceException(NSExceptionCode.RecordNotFound, new NSExceptionParams("d|r", domain, "email"));
+    }
+    return emailRecord;
+  }
+
+  @Override
+  String owner(String domain) throws NamingServiceException {
+   byte[] tokenId = this.tokenId(domain);
+   String owner = this.registryContract.getOwner(tokenId);
+    if (Utilities.isNull(owner)) {
+      throw new NamingServiceException(NSExceptionCode.UnregisteredDomain, new NSExceptionParams("d", domain));
+    }
+    return owner;
+  }
+
+  private Resolver getResolverContract(String domain) throws NamingServiceException {
+    String resolverAddress = getResolverAddress(domain);
+    if (Boolean.TRUE.equals(Utilities.isNull(resolverAddress))) {
+      // if owner is not found UnregisteredDomain is raised
+      this.owner(domain);
```

Comment:
@JohnnyJumper Please, throw `NSExceptionCode.UnregisteredDomain` explicitly. Using getter in this way can cause unpredicted side effects.